eth/filters/dropped_tx_subscription.go

```diff
@@ -2,23 +2,74 @@ package filters
 
 import (
 	"context"
+	"time"
 
-	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
 	"github.com/ethereum/go-ethereum/core"
 	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/internal/ethapi"
 	"github.com/ethereum/go-ethereum/rpc"
 )
 
 type dropNotification struct {
-	TxHash common.Hash `json:"txhash"`
-	Reason string      `json:"reason"`
+	// TxHash common.Hash `json:"txhash"`
+	Tx          *ethapi.RPCTransaction `json:"tx"`
+	Reason      string                 `json:"reason"`
+	Replacement *ethapi.RPCTransaction `json:"replacedby,omitempty"`
+	Peer        interface{}            `json:"peer,omitempty"`
+	Time        int64                  `json:"ts"`
 }
 
 type rejectNotification struct {
 	Tx     *types.Transaction
 	Reason string `json:"reason"`
 }
 
+// newRPCTransaction returns a transaction that will serialize to the RPC
+// representation, with the given location metadata set (if available).
+func newRPCPendingTransaction(tx *types.Transaction) *ethapi.RPCTransaction {
+	if tx == nil {
+		return nil
+	}
+	var signer types.Signer
+	if tx.Protected() {
+		signer = types.LatestSignerForChainID(tx.ChainId())
+	} else {
+		signer = types.HomesteadSigner{}
+	}
+	from, _ := types.Sender(signer, tx)
+	v, r, s := tx.RawSignatureValues()
+	result := &ethapi.RPCTransaction{
+		Type:     hexutil.Uint64(tx.Type()),
+		From:     from,
+		Gas:      hexutil.Uint64(tx.Gas()),
+		GasPrice: (*hexutil.Big)(tx.GasPrice()),
+		Hash:     tx.Hash(),
+		Input:    hexutil.Bytes(tx.Data()),
+		Nonce:    hexutil.Uint64(tx.Nonce()),
+		To:       tx.To(),
+		Value:    (*hexutil.Big)(tx.Value()),
+		V:        (*hexutil.Big)(v),
+		R:        (*hexutil.Big)(r),
+		S:        (*hexutil.Big)(s),
+	}
+	switch tx.Type() {
+	case types.AccessListTxType:
+		al := tx.AccessList()
+		result.Accesses = &al
+		result.ChainID = (*hexutil.Big)(tx.ChainId())
```

Comment:
Just double-checking:  are we 100% certain that `tx.ChainId` **always** returns a `*hexutil.Big`?  It can never be nil or return something else?

Same question for the type assertions below.

Reply:
no it returns 0 in worst case according to comments